src/utils/file.cpp

```diff
@@ -233,27 +276,139 @@ namespace file_util {
     return ret;
   }
 
+  /*
+   * Return the path without the last component (see dirname(3))
+   */
+  string dirname(const string& path) {
+    if (path.empty()) {
+      return ".";
+    }
+
+    auto start = path.size() - 1;
+    if (path[start] == '/' && start > 0) {
+      --start;
+    }
+
+    auto pos = path.rfind('/', start);
+    if (pos == string::npos) {
+      return ".";
+    }
+
+    if (pos == 0) { // path == "/" or path == "/directory"
+      return "/";
+    }
+
+    return path.substr(0, pos);
+  }
+
   /**
    * Path expansion
    */
-  const string expand(const string& path) {
-    string ret;
-    vector<string> p_exploded = string_util::split(path, '/');
-    for (auto& section : p_exploded) {
-      switch(section[0]) {
-        case '$':
-          section = env_util::get(section.substr(1));
-          break;
-        case '~':
-          section = env_util::get("HOME");
-          break;
+  string expand(const string& path) {
```

Comment:
Please document the code inside this function better as its quite a big chunk of code.

Reply:
Good point, I added more comments, let me know if that helps